subsys/pm/device_runtime.c

```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
 						  PM_DEVICE_STATE_SUSPEND,
 						  NULL, NULL);
 		}
-		ret = 0;
 		goto out;
 	}
 
-	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
+
+	if (!dev->pm->enable) {
+		ret = -ENOTSUP;
+		goto out_unlock;
+	}
+
+	if (target_state == PM_DEVICE_STATE_ACTIVE) {
+		dev->pm->usage++;
+		if (dev->pm->usage < 0) {
+			goto out_unlock;
+		}
+	} else {
+		dev->pm->usage--;
+		if (dev->pm->usage > 1) {
+			goto out_unlock;
+		}
+	}
+
 
 	/* Return in case of Async request */
 	if (pm_flags & PM_DEVICE_ASYNC) {
-		ret = 0;
-		goto out;
+		(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
+		goto out_unlock;
 	}
 
-	k_mutex_init(&request_mutex);
-	k_mutex_lock(&request_mutex, K_FOREVER);
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	pm_device_runtime_state_set(dev->pm);
```

Comment:
Setting the state should be based on the condition variable, so this line really needs to be after checking the condition.

Reply:
it does not to be fair. The state and the queue, that governs async ops, are protected by the mutex and not by the conditional variable. This code assumes that the device can do it asynchronously and because of that we need to wait after the operation (because only from the callback signal we will know that the operation finished). But, I saw the next comment and I got that this is coming from the assumption that devices do it synchronously.